x-pack/plugins/infra/server/utils/create_date_range_aggregation/create_date_range_aggregation.ts

```diff
@@ -0,0 +1,255 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import datemath from '@kbn/datemath';
+
+interface StringTimerange {
+  from: string;
+  to: string;
+}
+
+interface NumberTimerange {
+  from: number;
+  to: number;
+}
+
+type Timerange = StringTimerange | NumberTimerange;
+
+// Should we always drop partial ranges?
+type PartialRangeSetting = 'start' | 'end' | 'drop';
+type RangeSizeMode = 'minimum' | 'exact';
+
+interface Options {
+  timerange: Timerange;
+  targetRangeSize: string;
+  field?: string;
+  partialRangeSetting?: PartialRangeSetting;
+  targetGranularity?: number;
+}
+
+const defaults: Required<Omit<Options, 'timerange' | 'targetRangeSize'>> = {
+  field: '@timestamp',
+  partialRangeSetting: 'start',
+  targetGranularity: 100,
+};
+
+interface RangeSizeOptions {
+  mode: RangeSizeMode;
+  rangeSizeMs: number;
+  durationMs: number;
+  targetGranularity: number;
+}
+
+const unitToMilliseconds: Record<string, number> = {
+  s: 1000,
+  m: 60000,
+  h: 60000 * 60,
+  d: 60000 * 60 * 24,
+  w: 60000 * 60 * 24 * 7,
+  M: 60000 * 60 * 24 * 30,
```

Comment:
This unit is a bit odd since a month isn't exactly 30 days because "which month".
I feel like it would be better to just say "30d" instead, since "180d" isn't hard to write if you want "6m".